src/modules/esl-utils/dom/events/listener.ts

```diff
@@ -0,0 +1,148 @@
+
+const STORE = '__listeners';
+
+/** Event Listener instance, used as an 'inner' record to process subscriptions made by `EventUtils` */
+export class ESLEventListener implements ESLListenerDescriptor {
+  public readonly id: string;
+  public readonly event: string;
+  public readonly capture?: boolean;
+  public readonly selector?: string;
+  public readonly target?: string | EventTarget;
+
+  constructor(
+    public readonly $host: HTMLElement,
+    public readonly handler: ESLListenerHandler,
+    desc: ESLListenerDescriptor
+  ) {
+    desc.id = desc.id || sequentialUID('esl.event');
+    Object.assign(this, {capture: false}, desc);
+    this.handle = this.handle.bind(this);
+  }
+
+  /** @returns target element to listen */
+  @memoize()
+  public get $targets(): EventTarget[] {
+    if (this.target instanceof Document || this.target instanceof Element || this.target instanceof Window) {
+      return [this.target];
+    }
+    if (typeof this.target === 'string') {
+      return TraversingQuery.all(this.target, this.$host);
+    }
+    return [this.$host];
+  }
+
+  /**
+   * Checks if the passed criteria matches current event listener
+   * - `string` type criteria checked by the `event` name
+   * - `function` type criteria checked by the `handler` reference
+   * - `object` type criteria checked as descriptor using `isSimilar` comparer
+   * Note: `function` (handler) marcher has a priority over descriptor check, so handler function own properties will be ignored
+   */
+  public matches(desc?: ESLListenerCriteria): boolean {
+    if (typeof desc === 'string') return this.event === desc;
+    if (typeof desc === 'function') return this.handler === desc;
+    if (typeof desc === 'object') return isSimilar(this, desc);
+    return false;
+  }
+
+  /** Handles caught event (used as callback for low-level subscriptions) */
+  protected handle(e: Event): void {
+    if (this.isDelegatedTarget(e)) this.handler.call(this.$host, e, this);
+  }
+
+  /** Checks if the passed event can be handled by the curren event listener */
+  protected isDelegatedTarget(e: Event): boolean {
+    if (typeof this.selector !== 'string') return true;
+    const target = e.target;
+    const current = e.currentTarget;
+    if (!(target instanceof HTMLElement) || !(current instanceof HTMLElement)) return false;
+    return current.contains(target.closest(this.selector));
+  }
+
+  /** (Re-)Subscribes event listener instance */
+  public subscribe(): void {
+    this.unsubscribe();
+    memoize.clear(this, '$targets');
+    this.$targets.forEach((el: HTMLElement) => el.addEventListener(this.event, this.handle, this.capture));
+    ESLEventListener.get(this.$host).push(this);
+  }
+
+  /** Unsubscribes event listener instance */
+  public unsubscribe(): void {
+    this.$targets.forEach((el: HTMLElement) => el.removeEventListener(this.event, this.handle, this.capture));
```

Comment:
`el: EventTarget` to be more precise, but ts can define it itself, so you can omit type here as you do below with `listener`, or add a type to `listener`